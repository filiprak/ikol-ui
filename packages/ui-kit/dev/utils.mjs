import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
