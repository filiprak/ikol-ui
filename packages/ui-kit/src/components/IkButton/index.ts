import IkButton from './IkButton.vue';
export interface IkButton extends InstanceType<typeof IkButton> { };
export interface IkButtonProps extends Omit<InstanceType<typeof IkButton>, `$${string}`> { };
export { IkButton };
