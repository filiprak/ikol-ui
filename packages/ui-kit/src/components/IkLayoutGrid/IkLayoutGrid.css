.ik-layout-grid {
    position: relative;
    display: flex;
    flex: 1 1 auto;
    overflow: hidden;
    width: 100%;
}

.ik-layout-grid .ik-layout-grid-window:not(:last-of-type),
.ik-layout-grid-window--border {
    border-right: 1px solid var(--surface-darken-color);
}

.ik-layout-grid-item {
    display: flex;
    flex: 1 0 0;
    flex-direction: column;
    max-width: 100%;
    max-height: 100%;
}

.ik-layout-grid-item--row {
    flex-direction: row;
}

.ik-layout-grid-window {
    display: flex;
    align-items: stretch;
    flex-direction: column;
    overflow: auto;
    flex: 1 1 auto;
    max-width: 100%;
    max-height: 100%;
}

.ik-layout-grid-window--static {
    flex: 0 0 auto;
}

.ik-layout-grid-window__header {
    display: flex;
    flex: 0 0 auto;
    align-items: center;
    max-width: 100%;
    max-height: 100%;
    padding: 24px 25px 16px 25px;
    line-height: 32px;
}

.ik-layout-grid-window--dense .ik-layout-grid-window__header {
    padding: 15px;
    line-height: 24px;
}

.ik-layout-grid-window--no-header-pad .ik-layout-grid-window__header {
    padding: 0;
}

.ik-layout-grid-window__title {
    flex: 1 1 auto;
    font-size: 13px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    min-width: 0;
}

.ik-layout-grid-window__actions {
    flex: 0 0 auto;
    font-size: 18px;
}

.ik-layout-grid-window__actions .ik-button {
    padding: 4px 10px;
    border-radius: 20px;
}

.ik-layout-grid-window__content {
    position: relative;
    display: flex;
    flex-direction: column;
    align-items: stretch;
    flex: 1 1 auto;
    min-height: 0;
    overflow: hidden;
    max-width: 100%;
    max-height: 100%;
}

.ik-layout-grid-window__top,
.ik-layout-grid-window__bottom {
    flex: 0 0 auto;
    max-width: 100%;
}

.ik-layout-grid-window__action {
    display: inline-flex;
    align-items: center;
    font-size: 13px;
    margin-left: 2px;
    color: var(--on-background-color-3);
    cursor: pointer;
}

.ik-layout-grid-window__action:hover {
    color: var(--on-background-color-2);
}