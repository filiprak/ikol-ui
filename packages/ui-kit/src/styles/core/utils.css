/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
