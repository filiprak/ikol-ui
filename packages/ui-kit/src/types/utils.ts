import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
