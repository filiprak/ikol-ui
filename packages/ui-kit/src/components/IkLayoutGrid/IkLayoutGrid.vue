<template>
    <div class="ik-layout-grid">
        <slot></slot>
    </div>
</template>
<script setup lang="ts">
import '@ui/styles';
import './IkLayoutGrid.css';
</script>
