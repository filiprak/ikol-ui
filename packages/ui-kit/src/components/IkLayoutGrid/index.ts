import IkLayoutGrid from './IkLayoutGrid.vue';
import IkLayoutGridItem from './IkLayoutGridItem.vue';
import IkLayoutGridWindow from './IkLayoutGridWindow.vue';

export interface IkLayoutGrid extends InstanceType<typeof IkLayoutGrid> { };
export interface IkLayoutGridItem extends InstanceType<typeof IkLayoutGridItem> { };
export interface IkLayoutGridWindow extends InstanceType<typeof IkLayoutGridWindow> { };

export { IkLayoutGrid, IkLayoutGridItem, IkLayoutGridWindow };
