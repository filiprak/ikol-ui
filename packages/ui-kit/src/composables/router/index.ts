import type { InjectionKey, Ref } from 'vue';
import { inject } from 'vue';
import { getCurrentInstance, logError } from '@ui/utils/core';
import { IkRouterManager } from '@ui/composables/router/IkRouterManager';
import type { IkRouteManager } from '@ui/composables/router/IkRouteManager';
import type { RouteRecordRaw } from 'vue-router';
import { createRouter as _createRouter, createWebHashHistory } from 'vue-router';

export const ROUTER_SYMBOL: InjectionKey<IkRouterManager> = Symbol.for('ik-router');

export interface IkRouterOptions {
    routes: Readonly<RouteRecordRaw[]>
}

export function createRouter(options: IkRouterOptions): IkRouterManager {
    return new IkRouterManager(
        _createRouter({
            history: createWebHashHistory(),
            routes: options.routes,
        })
    );
}

export function useRouter(): IkRouterManager {
    let router;

    if (getCurrentInstance('useRouter')) {
        router = inject(ROUTER_SYMBOL);
    }

    if (!router) {
        router = createRouter({ routes: [] });
        logError('Injection "router" not found');
    }

    return router;
}

export function useRoute(): Ref<IkRouteManager> {
    let router;

    if (getCurrentInstance('useRoute')) {
        router = inject(ROUTER_SYMBOL);
    }

    if (!router) {
        router = createRouter({ routes: [] });
        logError('Injection "router" not found');
    }

    return router.route;
}
