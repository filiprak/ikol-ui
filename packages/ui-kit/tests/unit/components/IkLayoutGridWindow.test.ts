import type { RenderOptions } from '@ui/tests/tools/utils';
import { renderShallow } from '@ui/tests/tools/utils';
import { IkLayoutGridWindow } from '@ui/components/IkLayoutGrid';
import { IkButton } from '@ui/components/IkButton';
import { h } from 'vue';
import { getRouterMock } from '@ui/tests/mocks/router';

describe('IkLayoutGridWindow', () => {
    let router = getRouterMock();
    const getInstance = (options: RenderOptions<typeof IkLayoutGridWindow> = {}) => renderShallow(IkLayoutGridWindow, {
        router,
    }, options);

    beforeAll(() => {
        router = getRouterMock();
    });

    it('renders with correct default state', () => {
        const wrapper = getInstance();

        expect(wrapper.classes()).toStrictEqual([
            'ik-layout-grid-window',
        ]);

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({
            slots: {
                default: 'This is default slot...',
            },
        });

        expect(wrapper.text()).toContain('This is default slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with header slot', () => {
        const wrapper = getInstance({
            slots: {
                header: 'This is header slot...',
            },
        });

        expect(wrapper.text()).toContain('This is header slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with top slot', () => {
        const wrapper = getInstance({
            slots: {
                top: 'This is top slot...',
            },
        });

        expect(wrapper.text()).toContain('This is top slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with bottom slot', () => {
        const wrapper = getInstance({
            slots: {
                top: 'This is bottom slot...',
            },
        });

        expect(wrapper.text()).toContain('This is bottom slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with actions slot', () => {
        const wrapper = getInstance({
            slots: {
                actions: () => h('div', 'This is actions slot...'),
            },
        });

        expect(wrapper.text()).toContain('This is actions slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with header prop', () => {
        const wrapper = getInstance({
            props: {
                header: 'This is header slot...',
            },
        });

        expect(wrapper.text()).toContain('This is header slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with new action', () => {
        const wrapper = getInstance({
            attrs: {
                onIkNew: jest.fn(),
            },
        });

        expect(wrapper.findAllComponents(IkButton).at(0)?.isVisible()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('emits new event', () => {
        const newMock = jest.fn();
        const wrapper = getInstance({
            attrs: {
                onIkNew: newMock,
            },
            global: { stubs: { IkButton } },
        });

        expect(wrapper.findAllComponents(IkButton).at(0)?.isVisible()).toBe(true);

        wrapper.findAllComponents(IkButton).at(0)?.trigger('click');

        expect(newMock).toHaveBeenCalled();
    });

    it('renders with edit action', () => {
        const wrapper = getInstance({
            attrs: {
                onIkEdit: jest.fn(),
            },
        });

        expect(wrapper.findAllComponents(IkButton).at(0)?.isVisible()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('emits edit event', () => {
        const editMock = jest.fn();
        const wrapper = getInstance({
            attrs: {
                onIkEdit: editMock,
            },
            stubs: { IkButton },
        });

        expect(wrapper.findAllComponents(IkButton).at(0)?.isVisible()).toBe(true);

        wrapper.findAllComponents(IkButton).at(0)?.trigger('click');

        expect(editMock).toHaveBeenCalled();
    });

    it('renders with remove action', () => {
        const wrapper = getInstance({
            attrs: {
                onIkRemove: jest.fn(),
            },
        });

        expect(wrapper.findAllComponents(IkButton).at(0)?.isVisible()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('emits remove event', () => {
        const removeMock = jest.fn();
        const wrapper = getInstance({
            attrs: {
                onIkRemove: removeMock,
            },
            stubs: { IkButton },
        });

        expect(wrapper.findAllComponents(IkButton).at(0)?.isVisible()).toBe(true);

        wrapper.findAllComponents(IkButton).at(0)?.trigger('click');

        expect(removeMock).toHaveBeenCalled();
    });

    it('renders with clone action', () => {
        const wrapper = getInstance({
            attrs: {
                onIkClone: jest.fn(),
            },
        });

        expect(wrapper.findAllComponents(IkButton).at(0)?.isVisible()).toBe(true);
    });

    it('emits clone event', () => {
        const cloneMock = jest.fn();
        const wrapper = getInstance({
            attrs: {
                onIkClone: cloneMock,
            },
            stubs: { IkButton },
        });

        expect(wrapper.findAllComponents(IkButton).at(0)?.isVisible()).toBe(true);

        wrapper.findAllComponents(IkButton).at(0)?.trigger('click');

        expect(cloneMock).toHaveBeenCalled();
    });

    it('hides action buttons', () => {
        const wrapper = getInstance({
            props: {
                hide_new: true,
                hide_edit: true,
                hide_remove: true,
                hide_clone: true,
            },
            attrs: {
                onIkNew: jest.fn(),
                onIkEdit: jest.fn(),
                onIkRemove: jest.fn(),
                onIkClone: jest.fn(),
            },
        });

        expect(wrapper.findAllComponents(IkButton)).toHaveLength(0);
    });

    it('renders static', () => {
        const wrapper = getInstance({
            props: {
                static: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-layout-grid-window--static');
    });

    it('renders with border', () => {
        const wrapper = getInstance({
            props: {
                border_right: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-layout-grid-window--border');
    });

    it('renders dense', () => {
        const wrapper = getInstance({
            props: {
                dense: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-layout-grid-window--dense');
    });

    it('renders without header padding', () => {
        const wrapper = getInstance({
            props: {
                no_header_padding: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-layout-grid-window--no-header-pad');
    });
});
