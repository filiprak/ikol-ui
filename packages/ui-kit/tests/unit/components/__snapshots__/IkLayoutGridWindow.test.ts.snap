// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkLayoutGridWindow renders with actions slot 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__header"
  >
    <div
      class="ik-layout-grid-window__title"
    />
    <div
      class="ik-layout-grid-window__actions"
    >
      <div
        class="ik-layout-grid-window__action"
      >
        <div>
          This is actions slot...
        </div>
      </div>
    </div>
  </div>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;

exports[`IkLayoutGridWindow renders with bottom slot 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__top"
  >
    This is bottom slot...
  </div>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;

exports[`IkLayoutGridWindow renders with correct default state 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;

exports[`IkLayoutGridWindow renders with default slot 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__content"
  >
    This is default slot...
  </div>
</div>
`;

exports[`IkLayoutGridWindow renders with edit action 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__header"
  >
    <div
      class="ik-layout-grid-window__title"
    />
    <div
      class="ik-layout-grid-window__actions"
    >
      <div
        class="ik-layout-grid-window__action"
      >
        <ik-button-stub
          active="false"
          block="false"
          circle="false"
          design="default"
          disabled="false"
          fab="false"
          filled="false"
          flat="true"
          icon="pen:light"
          loading="false"
          outline="false"
          round="false"
          separate="false"
          skeleton="false"
          tag="button"
          type="button"
        />
      </div>
    </div>
  </div>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;

exports[`IkLayoutGridWindow renders with header prop 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__header"
  >
    <div
      class="ik-layout-grid-window__title"
    >
      This is header slot...
    </div>
    <div
      class="ik-layout-grid-window__actions"
    />
  </div>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;

exports[`IkLayoutGridWindow renders with header slot 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__header"
  >
    <div
      class="ik-layout-grid-window__title"
    >
      This is header slot...
    </div>
    <div
      class="ik-layout-grid-window__actions"
    />
  </div>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;

exports[`IkLayoutGridWindow renders with new action 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__header"
  >
    <div
      class="ik-layout-grid-window__title"
    />
    <div
      class="ik-layout-grid-window__actions"
    >
      <div
        class="ik-layout-grid-window__action"
      >
        <ik-button-stub
          active="false"
          block="false"
          circle="false"
          class="ik-mr-2"
          design="accent"
          disabled="false"
          fab="false"
          filled="true"
          flat="false"
          icon="plus"
          loading="false"
          outline="false"
          round="false"
          separate="false"
          skeleton="false"
          tag="button"
          type="button"
        />
      </div>
    </div>
  </div>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;

exports[`IkLayoutGridWindow renders with remove action 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__header"
  >
    <div
      class="ik-layout-grid-window__title"
    />
    <div
      class="ik-layout-grid-window__actions"
    >
      <div
        class="ik-layout-grid-window__action"
      >
        <ik-button-stub
          active="false"
          block="false"
          circle="false"
          design="default"
          disabled="false"
          fab="false"
          filled="false"
          flat="true"
          icon="trash-alt:light"
          loading="false"
          outline="false"
          round="false"
          separate="false"
          skeleton="false"
          tag="button"
          type="button"
        />
      </div>
    </div>
  </div>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;

exports[`IkLayoutGridWindow renders with top slot 1`] = `
<div
  class="ik-layout-grid-window"
>
  <div
    class="ik-layout-grid-window__top"
  >
    This is top slot...
  </div>
  <div
    class="ik-layout-grid-window__content"
  />
</div>
`;
